opengovsg/postmangovsg
# fix(email): Fix SendGrid fallback integration

## Problem

We currently use SendGrid as our email fallback. This causes the
following functions to fail:

1. Verifying custom from addresses (e.g. adding a new address, storing
   a campaign template with a custom address)
2. Sending emails with custom from addresses will fail
3. Receiving delivery report callbacks since the code currently parses the callbacks incorrectly

Closes #1021

## Solution

Since custom from addresses are not verified with SendGrid, we shouldn't use
them at all.

In the future, if we do intend to authenticate custom from addresses with SendGrid,
we will also have to rewire `verifyFromAddress()` to use SendGrid APIs.

**Features**:
1. Send all emails using the default address during the fallback.
2. Disable verification of custom from addresses during the fallback.
3. Disable storing of templates using custom from addresses.
4. Documented instructions to activate the email fallback on the incident response doc
5. Fix parsing of SendGrid webhook events

## Before & After Screenshots

<details>
  <summary>Click to show images</summary>

**AFTER**:
Attempting to verify a custom from address:
<img width="1534" alt="Screenshot 2021-03-10 at 9 07 24 AM" src="https://user-images.githubusercontent.com/4538946/110559938-126f1980-8180-11eb-8bb7-d5940fb542e5.png">

Attempting to save an email template using a custom from address:
<img width="1536" alt="Screenshot 2021-03-10 at 9 24 54 AM" src="https://user-images.githubusercontent.com/4538946/110561360-86aabc80-8182-11eb-8441-71900443f2da.png">

Creating a new email campaign:
<img width="1534" alt="Screenshot 2021-03-10 at 9 03 46 AM" src="https://user-images.githubusercontent.com/4538946/110559684-a42a5700-817f-11eb-8b8a-21053988263a.png">
</details>

## Tests

Test 1: all campaign emails should be sent using the default from address during fallback
- Store a campaign using a custom from address
- Set `EMAIL_FALLBACK_ACTIVATE` to `true` for the worker
- Send the campaign
- The campaign emails should be sent using the default from address

Test 2: verification of custom from addresses should be disabled during fallback
- Set `EMAIL_FALLBACK_ACTIVATE` to `true` for the backend
- Try to verify a custom from address
- The UI should display an error message like in the first screenshot above

Test 3: storing of templates with custom from addresses should be disabled during fallback
- Set `EMAIL_FALLBACK_ACTIVATE` to `true` for the backend
- Try to store a campaign using a custom from address
- The UI should display an error message like in the second screenshot above

**Test 4: SES callbacks should be received and parsed correctly**
- [x] Test sending via SES
  1. Deploy the branch on Travis with `STAGING_BRANCH=sendgrid-email-fallback` and `DEPLOY_WORKER=true`
  2. Send an email campaign (using these emails: https://docs.aws.amazon.com/ses/latest/DeveloperGuide/send-email-simulator.html)
  3. Ensure that the `status` and `error_code` columns of the`email_messages` are updated correctly]

**Test 5: SendGrid callbacks should be received and parsed correctly**
- [x] Test sending via SendGrid
  1. Set `EMAIL_FALLBACK_ACTIVATE=true` on Secrets Manager
  2. Prefix the Secrets Manager variables listed below with `BAK_`
  3. Add SendGrid-variants of the variables previously prefixed with `BAK_`
  4. Set the SendGrid event webhook HTTP POST URL to the correct callback server endpoint, with the correct username and password
  5. Deploy the branch on Travis with `STAGING_BRANCH=sendgrid-email-fallback` and `DEPLOY_WORKER=true`
  6. Send an email campaign (using a valid email and `bouncetest@tribulant.com`)
  2. Ensure that the `status` and `error_code` columns of the`email_messages` are updated correctly

## Deploy Notes

**New environment variables**:

- `EMAIL_FALLBACK_ACTIVATE`: Switch to true to use SendGrid as fallback for all emails. Ensure that the SMTP settings are properly configured as well. (worker and backend)

## Todo

- [x] Update the incident response doc to explain about the SendGrid fallback

```
BACKEND_SES_HOST
BACKEND_SES_USER
BACKEND_SES_PASS
WORKER_SES_HOST
WORKER_SES_USER
WORKER_SES_PASS
```
